11_notify/willshen8/pkg/puppy/syncstore.go

```diff
@@ -0,0 +1,75 @@
+package puppy
+
+import (
+	"strconv"
+	"sync"
+)
+
+type SyncStore struct {
+	syncStore sync.Map
+	nextID    uint32
+	mu        sync.Mutex
+}
+
+// NewMapStore initialise a new SyncStore
+func NewSyncStore() *SyncStore {
+	return &SyncStore{syncStore: sync.Map{}}
+}
+
+// CreatePuppy create a new puppy and store in mapStore.
+func (s *SyncStore) CreatePuppy(p *Puppy) (uint32, error) {
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	i, err := strconv.Atoi(p.Value)
```

Comment:
You do this same validation in four places. That really should be moved to a single place, especially as the mapstore version uses different error messages. I suggest you make the validation be a method on Puppy:
```
func (p* Puppy) Validate() error {
    if i, err := strconv.Atoi(p.Value); err != nil {
        return Errorf(ErrInvalidInput, "Puppy value not a number")
    } else if i < 0 {
        return Errorf(ErrInvalidInput, "Puppy value can't be negative")
    }
    return nil
}
```
and then use it in the store methods:
```
if err := p.Validate(); err != nil {
    return 0, err
}
```

Reply:
You are a champion in code simplicity and reuse, Cam.